### ISSUE TITLE
@AfterMapping method with @MappingTarget original not correctly identified when use Lombok @Builder

### ISSUE THREAD

**Issue by ClayZH** [mapstruct/mapstruct#3176]
### Expected behavior

```java
@Override
public ExampleRequestModel toRequestModel(ExampleV1Form form) {
    if ( form == null ) {
        return null;
    }

    ExampleRequestModel.ExampleRequestModelBuilder builder = ExampleRequestModel.builder();

    builder.example( form.getExample() );

    ExampleRequestModel exampleRequestModel = builder.build();

    handleOriginalAfterMapping( exampleRequestModel );

    return exampleRequestModel;
}
```

### Actual behavior

```java
@Override
public ExampleRequestModel toRequestModel(ExampleV1Form form) {
    if ( form == null ) {
        return null;
    }

    ExampleRequestModel.ExampleRequestModelBuilder exampleRequestModel = ExampleRequestModel.builder();

    exampleRequestModel.example( form.getExample() );

    return exampleRequestModel.build();
}
```

### Steps to reproduce the problem

Make 2 simple class, the mapping target use `@Builder` (lombok version):
```java
@Getter
@Setter
@NoArgsConstructor
public class ExampleV1Form {

    private String example;

}
```
```java
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class ExampleRequestModel {

    private String example;

}
```
then made a simple mapping:
```java
public abstract class ExampleMapper {

    public abstract ExampleRequestModel toRequestModel(ExampleV1Form form);

    @AfterMapping
    public void handleOriginalAfterMapping(@MappingTarget ExampleRequestModel requestModel) {
    	// Do something
    }
}
```
And what mapstruct generate is above [ Actual behavior ] part. The after mapping method have not been invoked.

And if i give a method with builder type, it can be invoke correctly:
```java
public abstract class ExampleMapper {

    public abstract ExampleRequestModel toRequestModel(ExampleV1Form form);

    @AfterMapping
    public void handleBuilderAfterMapping(@MappingTarget ExampleRequestModel.ExampleRequestModelBuilder builder) {
    	// Do something
    }
}

public class ExampleMapperImpl extends ExampleMapper {

    @Override
    public ExampleRequestModel toRequestModel(ExampleV1Form form) {
        if ( form == null ) {
            return null;
        }

        ExampleRequestModel.ExampleRequestModelBuilder exampleRequestModel = ExampleRequestModel.builder();

        exampleRequestModel.example( form.getExample() );

        handleBuilderAfterMapping( exampleRequestModel );

        return exampleRequestModel.build();
    }
}
```
 
What i expected is the after mapping method base on original type should be invoked correctly as if there is no `@Builder`
which like [ Expected behavior ] mentioned, because sometime handle logic is base on the final target not the middle one (builder, which can only be set not get in some version like lombok)

And if you can support both that's better:
```java
public abstract class ExampleMapper {

    public abstract ExampleRequestModel toRequestModel(ExampleV1Form form);

    @AfterMapping
    public void handleOriginalAfterMapping(@MappingTarget ExampleRequestModel requestModel) {
    	// Do something on original
    }

    @AfterMapping
    public void handleBuilderAfterMapping(@MappingTarget ExampleRequestModel.ExampleRequestModelBuilder builder) {
    	// Do something on builder
    }
}

public class ExampleMapperImpl extends ExampleMapper {

    @Override
    public ExampleRequestModel toRequestModel(ExampleV1Form form) {
        if ( form == null ) {
            return null;
        }

        ExampleRequestModel.ExampleRequestModelBuilder builder = ExampleRequestModel.builder();

        builder.example( form.getExample() );

        handleBuilderAfterMapping( builder );

        ExampleRequestModel exampleRequestModel = builder.build();

        handleOriginalAfterMapping( exampleRequestModel );

        return exampleRequestModel;
    }
}
```

### MapStruct Version

1.5.3.FINAL

**Comment by heowc** [comment]
> What i expected is the after mapping method base on original type should be invoked correctly as if there is no @Builder
> which like [ Expected behavior ] mentioned, because sometime handle logic is base on the final target not the middle one > (builder, which can only be set not get in some version like lombok)

These are different objects, but it would be great if we could make them feature 👍

**Comment by heowc** [comment]
Looking at the documentation, it seems that builder and AfterMapping can be used. 😎 

https://mapstruct.org/documentation/stable/reference/html/#customizing-mappings-with-before-and-after

**Comment by ClayZH** [comment]
> Looking at the documentation, it seems that builder and `@AfterMapping` can be used. 😎
> 
> https://mapstruct.org/documentation/stable/reference/html/#customizing-mappings-with-before-and-after

I have seen the Important Part:

> Important: when using a builder, the @AfterMapping annotated method must have the builder as @MappingTarget annotated parameter so that the method is able to modify the object going to be build. The build method is called when the @AfterMapping annotated method scope finishes. MapStruct will not call the @AfterMapping annotated method if the real target is used as @MappingTarget annotated parameter.

But i don't understand why the @AfterMapping method should not call the real target. Bases on builder, it can only be able to modify the build process (because some builder can only be set field, cannot get the field value from it). But most of the time, the after mapping method is handle logic after the real target builded, for example do some value validation, so if it only bases on builder, what after mapping can do is very limited.

**Comment by filiphr** [comment]
Your use case is valid @ClayZH. This is actually one of the points that is still not implemented. See #1454 as well.

What isn't working is `@AfterMapping` with `@MappingTarget` being the returned object.

Since we are a bit busy with different things, please let us know if there is someone from the community that would like to work on adding support for this. @ClayZH perhaps you'd like to give it a shot?

**Comment by venkatesh2090** [comment]
@filiphr, I spent some time trying to fix this use case here #3287. Still trying to think of test cases. Let me know if you think it's good.

**Comment by filiphr** [comment]
Closing this since it has been done by PR #3267
